server/container_create.go

```diff
@@ -798,19 +807,52 @@ func (s *Server) createSandboxContainer(ctx context.Context, containerID string,
 	specgen.HostSpecific = true
 	specgen.ClearProcessRlimits()
 
-	var readOnlyRootfs bool
+	readOnlyRootfs := s.config.ReadOnly
+
 	var privileged bool
 	if containerConfig.GetLinux().GetSecurityContext() != nil {
 		if containerConfig.GetLinux().GetSecurityContext().Privileged {
 			privileged = true
 		}
-
 		if containerConfig.GetLinux().GetSecurityContext().ReadonlyRootfs {
 			readOnlyRootfs = true
-			specgen.SetRootReadonly(true)
 		}
 	}
+	specgen.SetRootReadonly(readOnlyRootfs)
 
+	if s.config.ReadOnly {
+		options := []string{"rw", "noexec", "nosuid", "nodev", "tmpcopyup"}
```

Comment:
`tmpcopyup` is a runc extention (/opencontainers/runc#845) that is not part of the OCI spec.  More on this sort of thing in #1260, but we may want to document the out-of-spec extensions that CRI-O expects in a runtime if folks expect to be able to plug in an alternative, runtime-spec-compliant runtime.  For example, I know @giuseppe has recently been working on CRI-O compat for [crun][1], and crun does not currently support this extention.

[1]: https://github.com/giuseppe/crun

Reply:
An alternative to a `tmpfs` with `tmpcopyup` would be an [`overlay` mount][1].  Are there reasons that approach does not work (ping @mrunalp)?  If it's just “we want to support kernels that do not include `overlay` support”, perhaps we can check `/proc/filesystems` at `crio` start-time and use that to switch on our approach here.  That would reduce the `tmpcopyup` runtime requirement to systems which did not include `overlay` support.

[1]: https://www.kernel.org/doc/Documentation/filesystems/overlayfs.txt

Reply:
> An alternative to a tmpfs with tmpcopyup would be an overlay mount.

Something else to check for with this approach is whether the user is priveleged or the kernel support unprivileged `overlay` mounts.  More on that in opencontainers/runc#1671 and opencontainers/runc#1688, which point out that currently only Ubuntu (and maybe a few other distros?) allow for overlay mounts inside an unprivileged user namespace.

Reply:
Added a comment about this, linked this discussion there as well.